fix(RHTAPREL-652): add a check if mapped false push snapshot stop

happybhati opened the pull request:
- push-snapshot shouldn't run if mapped = false

happybhati committed:
fix(RHTAPREL-652): add a check if mapped false push snapshot stop

- Added `when` clause to `push-snapshot` task in the pipeline
  to ensure it only executes when  the `apply-mapping` task
  indicates that mapping was successful.

- Updated `apply-mapping` task to replace `echo` with
  `printf` for storing results. Using `echo` resulted in
  appending `true\n` to the output, which caused discrepancies
  with the when expression in the pipeline.

Signed-off-by: Happy Bhati <hbhati@redhat.com>

johnbieren commented:
I think this can be done a lot more cleanly using the `when` clause in the pipeline definitions

happybhati commented:
/test release-pipelines-e2e-suite

johnbieren reviewed (commented):
LGTM but same question about pipeline execution I posted here https://github.com/redhat-appstudio/release-service-catalog/pull/385#pullrequestreview-1949652129
Once the answer is figured out there, good to approve this one

happybhati commented:
/test release-pipelines-e2e-suite

happybhati commented:
/test release-pipelines-e2e-suite

happybhati commented:
/test release-pipelines-e2e-suite

happybhati commented:
/test release-pipelines-e2e-suite

happybhati commented:
/test release-pipelines-e2e-suite

jinqi7 commented:
/retest

jinqi7 commented:
/retest

jinqi7 commented:
/test release-pipelines-e2e-suite

happybhati commented:
/retest

happybhati commented:
/retest

happybhati commented:
/retest

happybhati commented:
/retest

happybhati commented:
/retest

jinqi7 commented:
/retest

jinqi7 commented:
/retest

jinqi7 commented:
/retest

happybhati commented:
> I found that the push-snapshot was skipped and the following tasks were also skipped in "push_to_external_registry" and "rh_push_to_external_registry"  pipelineRun except clean-up task. Then some checks in e2e-tests( such as if the images exist)  after the managed PipelineRun could not pass. It's not expected, right?

@jinqi7  If I am not wrong if mapped is false image will not be pushed and test needs to be modified to meet this change.

johnbieren commented:
> > I found that the push-snapshot was skipped and the following tasks were also skipped in "push_to_external_registry" and "rh_push_to_external_registry"  pipelineRun except clean-up task. Then some checks in e2e-tests( such as if the images exist)  after the managed PipelineRun could not pass. It's not expected, right?
> 
> @jinqi7 If I am not wrong if mapped is false image will not be pushed and test needs to be modified to meet this change.

If the mapping is wrong and no mapping takes place, the images aren't pushed so e2e should fail. So I would say yes, it is expected @jinqi7

johnbieren reviewed (commented):
So the tasks that rely on the push-snapshot.results.commonTags won't run if the push-snapshot task is skipped as the result won't be populated? That is why we don't need to add the when to as many tasks?

happybhati commented:
> So the tasks that rely on the push-snapshot.results.commonTags won't run if the push-snapshot task is skipped as the result won't be populated? That is why we don't need to add the when to as many tasks?

Exactly, I added 'when' only to those tasks which do not depend on push-snapshot and we need them to be skipped.

jinqi7 commented:
> > So the tasks that rely on the push-snapshot.results.commonTags won't run if the push-snapshot task is skipped as the result won't be populated? That is why we don't need to add the when to as many tasks?
> 
> Exactly, I added 'when' only to those tasks which do not depend on push-snapshot and we need them to be skipped.

If it's expected, the e2e-tests need to change accordingly.

johnbieren commented:
> > > So the tasks that rely on the push-snapshot.results.commonTags won't run if the push-snapshot task is skipped as the result won't be populated? That is why we don't need to add the when to as many tasks?
> > 
> > 
> > Exactly, I added 'when' only to those tasks which do not depend on push-snapshot and we need them to be skipped.
> 
> If it's expected, the e2e-tests need to change accordingly. I submit the PR [redhat-appstudio/e2e-tests#1115](https://github.com/redhat-appstudio/e2e-tests/pull/1115). But I am not sure if the "images.floatingTags" need to be added to the RPA.

Jing, this PR is to guard against misconfiguration. If mapped is false, the user set things up incorrectly to use this pipeline. So, with Happy's work here, other things will not happen that would break anyways. There is no e2e change for this

jinqi7 commented:
> > > > So the tasks that rely on the push-snapshot.results.commonTags won't run if the push-snapshot task is skipped as the result won't be populated? That is why we don't need to add the when to as many tasks?
> > > 
> > > 
> > > Exactly, I added 'when' only to those tasks which do not depend on push-snapshot and we need them to be skipped.
> > 
> > 
> > If it's expected, the e2e-tests need to change accordingly. I submit the PR [redhat-appstudio/e2e-tests#1115](https://github.com/redhat-appstudio/e2e-tests/pull/1115). But I am not sure if the "images.floatingTags" need to be added to the RPA.
> 
> Jing, this PR is to guard against misconfiguration. If mapped is false, the user set things up incorrectly to use this pipeline. So, with Happy's work here, other things will not happen that would break anyways. There is no e2e change for this

The current e2e test cases failed with this PR for the push-snapshot task and following tasks were skipped,  but the release pipelineRun passed finally. The release pipelineRun should not pass or not? Release should success or not?  In the real world, is "floatingTags" need to set for the "push-to-external-registry" / rh-push-to-external-registry pipelines ?

johnbieren commented:
> > > > > So the tasks that rely on the push-snapshot.results.commonTags won't run if the push-snapshot task is skipped as the result won't be populated? That is why we don't need to add the when to as many tasks?
> > > > 
> > > > 
> > > > Exactly, I added 'when' only to those tasks which do not depend on push-snapshot and we need them to be skipped.
> > > 
> > > 
> > > If it's expected, the e2e-tests need to change accordingly. I submit the PR [redhat-appstudio/e2e-tests#1115](https://github.com/redhat-appstudio/e2e-tests/pull/1115). But I am not sure if the "images.floatingTags" need to be added to the RPA.
> > 
> > 
> > Jing, this PR is to guard against misconfiguration. If mapped is false, the user set things up incorrectly to use this pipeline. So, with Happy's work here, other things will not happen that would break anyways. There is no e2e change for this
> 
> The current e2e test cases failed with this PR for the push-snapshot task and following tasks were skipped, but the release pipelineRun passed finally. The release pipelineRun should pass or not? Release should success or not? In the real world, is "floatingTags" need to be set for the "push-to-external-registry" / rh-push-to-external-registry pipelines ?

This PR should not affect the e2e-tests at all. The e2e-tests should map the components

jinqi7 commented:
/retest

jinqi7 commented on pipelines/push-to-external-registry/push-to-external-registry.yaml:
```diff
@@ -190,6 +190,10 @@ spec:
       runAfter:
         - apply-mapping
     - name: push-snapshot
+      when:
```
"When" clauses don't work as expected from my testing. Please use `printf "true" |  tee $(results.mapped.path)` instead of the line of  `echo "true" | tee $(results.mapped.path)`  in "apply-mapping" task to make them work well.  I tried that and it worked.  I referred the  [link](https://stackoverflow.com/questions/76425780/tekton-debugging-why-a-when-clause-is-evaluated-as-false). FYI.

jinqi7 commented:
> > > > > > So the tasks that rely on the push-snapshot.results.commonTags won't run if the push-snapshot task is skipped as the result won't be populated? That is why we don't need to add the when to as many tasks?
> > > > > 
> > > > > 
> > > > > Exactly, I added 'when' only to those tasks which do not depend on push-snapshot and we need them to be skipped.
> > > > 
> > > > 
> > > > If it's expected, the e2e-tests need to change accordingly. I submit the PR [redhat-appstudio/e2e-tests#1115](https://github.com/redhat-appstudio/e2e-tests/pull/1115). But I am not sure if the "images.floatingTags" need to be added to the RPA.
> > > 
> > > 
> > > Jing, this PR is to guard against misconfiguration. If mapped is false, the user set things up incorrectly to use this pipeline. So, with Happy's work here, other things will not happen that would break anyways. There is no e2e change for this
> > 
> > 
> > The current e2e test cases failed with this PR for the push-snapshot task and following tasks were skipped, but the release pipelineRun passed finally. The release pipelineRun should pass or not? Release should success or not? In the real world, is "floatingTags" need to be set for the "push-to-external-registry" / rh-push-to-external-registry pipelines ?
> 
> This PR should not affect the e2e-tests at all. The e2e-tests should map the components

I closed the PR in e2e-tests since there is no change needed there. In this PR,  "When" clauses don't work as expected and I added a comment for that.

johnbieren commented:
> > > > > > > So the tasks that rely on the push-snapshot.results.commonTags won't run if the push-snapshot task is skipped as the result won't be populated? That is why we don't need to add the when to as many tasks?
> > > > > > 
> > > > > > 
> > > > > > Exactly, I added 'when' only to those tasks which do not depend on push-snapshot and we need them to be skipped.
> > > > > 
> > > > > 
> > > > > If it's expected, the e2e-tests need to change accordingly. I submit the PR [redhat-appstudio/e2e-tests#1115](https://github.com/redhat-appstudio/e2e-tests/pull/1115). But I am not sure if the "images.floatingTags" need to be added to the RPA.
> > > > 
> > > > 
> > > > Jing, this PR is to guard against misconfiguration. If mapped is false, the user set things up incorrectly to use this pipeline. So, with Happy's work here, other things will not happen that would break anyways. There is no e2e change for this
> > > 
> > > 
> > > The current e2e test cases failed with this PR for the push-snapshot task and following tasks were skipped, but the release pipelineRun passed finally. The release pipelineRun should pass or not? Release should success or not? In the real world, is "floatingTags" need to be set for the "push-to-external-registry" / rh-push-to-external-registry pipelines ?
> > 
> > 
> > This PR should not affect the e2e-tests at all. The e2e-tests should map the components
> 
> I closed the PR in e2e-tests since there is no change needed there. In this PR, "When" clauses don't work as expected and I added a comment for that.

Thanks, I assumed Happy had tested this. @happybhati you didn't test this before marking it ready for review?

happybhati commented:
> > > > > > > > So the tasks that rely on the push-snapshot.results.commonTags won't run if the push-snapshot task is skipped as the result won't be populated? That is why we don't need to add the when to as many tasks?
> 
> > > > > > > 
> 
> > > > > > > 
> 
> > > > > > > Exactly, I added 'when' only to those tasks which do not depend on push-snapshot and we need them to be skipped.
> 
> > > > > > 
> 
> > > > > > 
> 
> > > > > > If it's expected, the e2e-tests need to change accordingly. I submit the PR [redhat-appstudio/e2e-tests#1115](https://github.com/redhat-appstudio/e2e-tests/pull/1115). But I am not sure if the "images.floatingTags" need to be added to the RPA.
> 
> > > > > 
> 
> > > > > 
> 
> > > > > Jing, this PR is to guard against misconfiguration. If mapped is false, the user set things up incorrectly to use this pipeline. So, with Happy's work here, other things will not happen that would break anyways. There is no e2e change for this
> 
> > > > 
> 
> > > > 
> 
> > > > The current e2e test cases failed with this PR for the push-snapshot task and following tasks were skipped, but the release pipelineRun passed finally. The release pipelineRun should pass or not? Release should success or not? In the real world, is "floatingTags" need to be set for the "push-to-external-registry" / rh-push-to-external-registry pipelines ?
> 
> > > 
> 
> > > 
> 
> > > This PR should not affect the e2e-tests at all. The e2e-tests should map the components
> 
> > 
> 
> > I closed the PR in e2e-tests since there is no change needed there. In this PR, "When" clauses don't work as expected and I added a comment for that.
> 
> 
> 
> Thanks, I assumed Happy had tested this. @happybhati you didn't test this before marking it ready for review?

Strange, I remember testing both changes and wondering how I did not catch this behavior at first place, I will do my testing again and fix accordingly. Thanks a lot, @jinqi7 for your findings **++**.

johnbieren commented on tasks/apply-mapping/apply-mapping.yaml:
```diff
@@ -55,15 +55,15 @@ spec:
 
         if [ ! -f "${RPA_FILE}" ] ; then
             echo "No ReleasePlanAdmission file was found."
-            echo "false" | tee $(results.mapped.path)
```
I assume the issue was the new line character, so `echo -n` probably would've fixed it too, but I am fine with printf

happybhati replied:
ahh, thanks, good to know!
